scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -0,0 +1,324 @@
+
+    case class Expression(
+        expressionPosition: ExpressionPosition,
+        anchorPosition: AnchorPosition,
+        rules: String
+    ) extends Escape
+
+    case class ExpressionLine(
+        expressionPosition: ExpressionPosition,
+        anchorPosition: AnchorPosition,
+        rules: String
+    ) extends Escape
+  }
+
+  private sealed abstract class Toogle
+  private object Toogle {
+    case object Disable extends Toogle
+    case object Enable extends Toogle
+  }
+
+  private val unusedScalafixSupressionId = "UnusedScalafixSupression"
+
+  private[scalafix] def unusedScalafixSupression(
+      idSuffix: String,
+      explain: String) =
+    LintCategory.warning(unusedScalafixSupressionId + idSuffix, explain)
+
+  private val unusedEnableWarning =
+    unusedScalafixSupression(
+      "Enable",
+      "This comment would enable a rule that was not disabled (eg: typo in the rules)"
+    )
+
+  private val FilterDisable = "\\s?scalafix:off\\s?(.*)".r
+  private val FilterEnable = "\\s?scalafix:on\\s?(.*)".r
+  private val FilterExpression = "\\s?scalafix:ok\\s?(.*)".r
+
+  private def splitRules(rules: String): Set[String] =
+    rules.split("\\s+").toSet
+
+  def apply(tree: Tree, associatedComments: AssociatedComments): EscapeHatch = {
+    val escapes = List.newBuilder[Escape]
+    val unusedEnable = List.newBuilder[LintMessage]
+    val visitedFilterExpression = mutable.Set.empty[AnchorPosition]
+
+    var currentlyDisabledRules = Set.empty[String]
+    def trackUnusedRules(
+        rules: String,
+        toogle: Toogle,
+        anchorPosition: AnchorPosition): Unit = {
+      val toogledRules = splitRules(rules)
+
+      toogle match {
+        case Toogle.Disable =>
+          currentlyDisabledRules = currentlyDisabledRules ++ splitRules(rules)
+
+        case Toogle.Enable =>
+          val enabledRules = splitRules(rules)
+          val enabledNotDisabled = enabledRules -- currentlyDisabledRules
+
+          enabledNotDisabled.foreach(
+            rule => unusedEnable += unusedEnableWarning.at(anchorPosition.value)
+          )
+      }
+    }
+
+    tree.foreach { t =>
+      associatedComments.trailing(t).foreach {
+        // matches simple expressions
+        //
+        // val a = (
+        //   1,
+        //   2
+        // ) // scalafix:ok RuleA
+        //
+        case comment @ Token.Comment(FilterExpression(rules)) =>
+          val anchorPosition = AnchorPosition(comment.pos)
+          if (!visitedFilterExpression.contains(anchorPosition)) {
+            escapes += Escape.Expression(
+              ExpressionPosition(t.pos),
+              anchorPosition,
+              rules
+            )
+            visitedFilterExpression += anchorPosition
+          }
+
+        case _ => ()
+      }
+    }
+
+    tree.tokens.foreach {
+      // matches off anchors
+      //
+      // // scalafix:off RuleA
+      // ...
+      //
+      case comment @ Token.Comment(FilterDisable(rules)) =>
```

Comment:
Can we match onces on `Token.Comment` and move the `FilterDisable` match inside the case body?